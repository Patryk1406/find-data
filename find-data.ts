const checkIfHaveCommonProperty = <T>(
  obj1: Record<string, T>,
  obj2: Record<string, T>,
): boolean => {
  for (const key of Object.keys(obj1)) {
    if (obj1[key] === obj2[key]) return true;
  }
  return false;
};

const checkIfThereIsAConflict = <T>(
  obj1: Record<string, T>,
  obj2: Record<string, T>,
): boolean => {
  for (const key of Object.keys(obj1)) {
    if (obj2[key] !== undefined && obj1[key] !== obj2[key]) return true;
  }
  return false;
};

export const findData = <T extends Record<string, unknown>, K extends Record<string, unknown>>(
  array: T[], array2: K[]): (T | K)[] => {
  const combined = [...array, ...array2];
  return combined.reduce((acc: (T | K)[], curr: T | K) => {
    const helper = <Y>(arr: Record<string, Y>[], obj: Record<string, Y>): boolean => {
      for (const element of arr) {
        if (checkIfHaveCommonProperty(element, obj) && !checkIfThereIsAConflict(element, obj)) {
          Object.assign(element, obj);
          Object.assign(obj, element);
          return true;
        }
        for (const value of Object.values(element)) {
          if (!value) break;
          if (Array.isArray(value)) {
            helper(value, obj);
            break;
          }
          if (typeof value === 'object') {
            if (checkIfHaveCommonProperty(value as Y & Record<string, any>, obj)
              && !checkIfThereIsAConflict(value as Y & Record<string, any>, obj)) {
              Object.assign(value, obj);
              Object.assign(obj, value);
              break;
            }
            helper(Object.values(value), obj);
          }
        }
      }
      return false;
    };
    const isMerged = helper(acc, curr);
    if (!isMerged) acc[acc.length] = curr;
    return acc;
  }, []);
};
