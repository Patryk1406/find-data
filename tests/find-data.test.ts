import { findData } from '../find-data';

describe('findData function test suite', () => {
  test('combined not embedded objects', () => {
    const input1: Record<string, string | number | Record<string, string | number>>[] = [
      {
        id: 1,
        name: 'Grzegorz',
      },
      {
        name: 'Grzegorz',
        surname: 'NieGrzegorz',
        animal: {
          name: 'Czarny',
          age: 1,
        },
      }];
    const input2: Record<string, string | number>[] = [
      {
        id: 1,
        city: 'Gliwice',
      },
    ];

    const output: (Record<string, string | number> |
    Record<string, string | number | Record<string, string | number>>)[] = findData(
      input1,
      input2,
    );

    expect(output)
      .toMatchSnapshot();
  });
  test('combined embedded objects and not embedded objects', () => {
    const input1: Record<string, string | number | Record<string, string | number>>[] = [
      {
        id: 1,
        name: 'Grzegorz',
      },
      {
        name: 'Grzegorz',
        surname: 'NieGrzegorz',
        animal: {
          name: 'Czarny',
          age: 1,
        },
      },
      {
        id: 1,
        city: 'Gliwice',
      },
      {
        age: 1,
        color: 'Black',
      }];
    const input2: Record<string, string | { id: number; name: string; }[]>[] = [
      {
        friends: [
          {
            id: 1241,
            name: 'Adam',
          }],
      },
      {
        name: 'Adam',
        lastName: 'NieAdam',
      }];

    const output: (Record<string, string | number |
    Record<string, string | number>> |
    Record<string, string | { id: number, name: string }[]>)[] = findData(input1, input2);

    expect(output)
      .toMatchSnapshot();
  });
});
