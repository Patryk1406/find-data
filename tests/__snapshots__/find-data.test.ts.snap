// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findData function test suite combined embedded objects and not embedded objects 1`] = `
Array [
  Object {
    "animal": Object {
      "age": 1,
      "color": "Black",
      "name": "Czarny",
    },
    "city": "Gliwice",
    "id": 1,
    "name": "Grzegorz",
    "surname": "NieGrzegorz",
  },
  Object {
    "age": 1,
    "color": "Black",
    "name": "Czarny",
  },
  Object {
    "friends": Array [
      Object {
        "id": 1241,
        "lastName": "NieAdam",
        "name": "Adam",
      },
    ],
  },
  Object {
    "id": 1241,
    "lastName": "NieAdam",
    "name": "Adam",
  },
]
`;

exports[`findData function test suite combined not embedded objects 1`] = `
Array [
  Object {
    "animal": Object {
      "age": 1,
      "name": "Czarny",
    },
    "city": "Gliwice",
    "id": 1,
    "name": "Grzegorz",
    "surname": "NieGrzegorz",
  },
]
`;
